code/modules/migrations/migrations_controller.dm

```diff
@@ -0,0 +1,114 @@
+/datum/migration_controller
+	var/list/db_states[0]
+	var/list/packages[0]
+
+	var/TABLE_NAME = "_migrations"
+	var/id = ""
+
+/datum/migration_controller/proc/setup()
+	return FALSE
+/datum/migration_controller/proc/createMigrationTable()
+	return FALSE
+
+/datum/migration_controller/New()
+	if(!setup())
+		world.log << "\[Migrations] ([id]): Setup() returned false, will not run migrations for this DBMS."
+	else
+		if(!hasTable(TABLE_NAME))
+			world.log << "\[Migrations] ([id]): Creating [format_table_name(TABLE_NAME)]"
+			createMigrationTable()
+
+		for(var/list/row in query("SELECT pkgID, version FROM [format_table_name(TABLE_NAME)]"))
+			if(id=="mysql")
+				db_states[row[1]] = text2num(row[2])
+			else
+				db_states[row["pkgID"]] = text2num(row["version"])
+
+		var/list/newpacks[0]
+		for(var/mtype in typesof(/datum/migration)-list(/datum/migration))
+			var/datum/migration/M = new mtype(src)
+			if(M.package == "" || M.name == "" || M.dbms != id)
```

Comment:
reduce errors in the future

ps. not entirely sure this will operate correctly, I'm pretty sure but not 100% so just check that initial won't grab the initial of the child's type instead of the parent (datum/migration) -- it would be unexpected but I don't know for sure

```suggestion
			if(M.package == initial(M.package) || M.name == initial(M.name) || M.dbms != id)
```